src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotCallDangerousMethodsInDeserializationTests.cs

```diff
@@ -0,0 +1,1817 @@
+Imports System
+Imports System.IO
+Imports System.Runtime.Serialization
+
+Namespace TestNamespace
+    <Serializable()> _
+    Class TestClass
+        Private member As String
+        
+        <OnDeserialized()>
+        Sub OnDeserializedMethod(ByVal context As StreamingContext)
+            Dim obj As New TestClass()
+            Dim bytes(9) As Byte
+            File.WriteAllBytes(""C:\\"", bytes)
+        End Sub
+
+        Sub TestMethod()
+            Dim bytes(9) As Byte
+            File.WriteAllBytes(""C:\\"", bytes)
+        End Sub
+    End Class
+End Namespace",
+            GetBasicResultAt(12, 13, DoNotCallDangerousMethodsInDeserialization.Rule, "TestClass", "OnDeserializedMethod", "WriteAllBytes"));
+        }
+
+        [Fact]
+        public void TestOnDeserializedMultiMediateInvocationsDiagnostic()
+        {
+            VerifyCSharp(@"
+using System;
+using System.IO;
+using System.Runtime.Serialization;
+
+[Serializable()]
+public class TestClass
+{
+    private string member;
+
+    [OnDeserialized()]
+    internal void OnDeserializedMethod(StreamingContext context)
+    {
+        var obj = new TestClass();
+        var count = 2;
+        obj.TestMethod(count);
+    }
+    
+    private void TestMethod(int count)
+    {
+        byte[] bytes = new byte[] {0x20, 0x20, 0x20};
+        File.WriteAllBytes(""C:\\"", bytes);
+
+        if(count != 0)
+        {
+            var obj = new TestClass();
+            obj.TestMethod(--count);
+        }
+    }
+}",
+            GetCSharpResultAt(12, 19, DoNotCallDangerousMethodsInDeserialization.Rule, "TestClass", "OnDeserializedMethod", "WriteAllBytes"));
+
+            VerifyBasic(@"
+Imports System
+Imports System.IO
+Imports System.Runtime.Serialization
+
+Namespace TestNamespace
+    <Serializable()> _
+    Class TestClass
+        Private member As String
+        
+        <OnDeserialized()>
+        Sub OnDeserializedMethod(ByVal context As StreamingContext)
+            Dim obj As New TestClass()
+            obj.TestMethod(2)
+        End Sub
+
+        Sub TestMethod(ByVal count As Integer)
+            Dim bytes(9) As Byte
+            File.WriteAllBytes(""C:\\"", bytes)
+
+            If count <> 0
+                Dim obj As New TestClass()
+                count = count - 1
+                obj.TestMethod(count)
+            End If
+        End Sub
+    End Class
+End Namespace",
+            GetBasicResultAt(12, 13, DoNotCallDangerousMethodsInDeserialization.Rule, "TestClass", "OnDeserializedMethod", "WriteAllBytes"));
+        }
+
+        [Fact]
+        public void TestOnDeserializationWriteAllBytesDiagnostic()
+        {
+            VerifyCSharp(@"
+using System;
+using System.IO;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Text;
+
+[Serializable()]
+public class TestClass : IDeserializationCallback 
+{
+    private string member;
+
+    void IDeserializationCallback.OnDeserialization(Object sender) 
```

Comment:
Have any tests where `IDeserializationCallback.OnDeserialization` isn't explicitly implemented (e.g. in C#, just `void OnDeserialization(object sender)`)?

Reply:
This interface can only be implemented explicitly. But your advice is good, I add one, and it doesn't report any diagnostic as expected.

Reply:
What about something like this?

```csharp
using System;
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;

namespace TestDeserialization
{
    [Serializable]
    public class MyData : IDeserializationCallback
    {
        public int MyInt { get; set; }

        public void OnDeserialization(object sender)
        {
            Console.WriteLine("In OnDeserialization()");
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            BinaryFormatter f = new BinaryFormatter();
            using (MemoryStream ms = new MemoryStream())
            {
                f.Serialize(ms, new MyData() { MyInt = 1 });

                ms.Position = 0;
                MyData d = (MyData) f.Deserialize(ms);
            }
        }
    }
}
```